### ISSUE TITLE
Random Exceptions in HttpProcessor's Process method

### ISSUE THREAD

**Issue by petrroll** [JohnnyCrazy/SpotifyAPI-NET#98]
Hi, I'm getting random `IOExceptions ` in `HttpProcessor`'s `Process` method, specifically on Flush() call. It seems that something sometimes causes the given `TcpClient` to close without any warning which consequentially causes  `IOExceptions` on `OutputStream`'s `Flush` method. 

I couldn't reliably reproduce the error as it happens only sometimes but just connecting with Spotify web API from the example app repeatedly usually results in the bug showing up after 2-5 tries. 

I tried to debug the issue but didn't progress at all.

**Comment by JohnnyCrazy** [comment]
Hi @petrroll 

Yes, the HTTPServer is kinda confusing. It was based on a very simple and rudimentary stackoverflow answer and I didn't change much. I will have a look at it through the next days, maybe a rewrite is indeed a good decision

**Comment by erinxocon** [comment]
If the server is rewritten with support for the .net core framework in mind then the web API should be supported on multiple platforms. That's where I ran into trouble in my attempt to port the API. Just saying 🙂

**Comment by JohnnyCrazy** [comment]
@erinxocon 
Just did a quick research, maybe we can utilize https://github.com/aspnet/KestrelHttpServer, but this would add a strict dependency to .net core which I would like to avoid

**Comment by erinxocon** [comment]
I might be completely off on this but dotnet core projects allow you to target multiple frameworks.  Would it be possible to detect the runtime and choose what to use?  I know for a project I'm working on that uses .NET core, I can use Kestral or IIS.

**Comment by petrroll** [comment]
I actually wanted to start working on the rewrite myself but unfortunately got quite sick. I should be both ok and have some time next week, though. So if you don't start by then I'll be happy to take a look at it :).

@JohnnyCrazy @erinxocon judging by [nugget page listing](https://www.nuget.org/packages/Kestrel) and some quick googling it seems that Kestrel is happy to use either Core (DNXCore 5.0) or the full framework (DNX 4.5.1).

Anyway, isn't using Kestrel for such a simple task a bit overkill? I frankly don't know, never used Kestrel (actually my only knowledge of it comes from two NDC talks).

**Comment by JohnnyCrazy** [comment]
@petrroll Yea, could be a bit overkill considering we just need to handle 3 different auth methods and simple stuff. I also want to keep dependencies low, so maybe we can rely on our own implementation. 

I still didn't have time to look at .NET core in detail, but I guess we will need another project to build for it?

**Comment by erinxocon** [comment]
We can build on my forked version.  I have a [branch](https://github.com/erinxocon/SpotifyAPI-NET/tree/dotnetcore) for .NET core work.  What three auth methods do we need to implement, I can also look at this this weekend.  Needs a bit of organization, also it's a vscode project but could be a Visual Studio project if need be.

**Comment by petrroll** [comment]
@erinxocon 
If do that please rebase your fork first so that we don't loose progress made here.

BTW: As much as I'm for using the newest tech possible I'd probably advocate for waiting a bit before porting it to .NET core. There're still things that will change in regards to .NET Core toolkit (such as abandoning project.json) so some work might be done in vain.

I'm, however, for rewriting the WebServer in a way that will make the following port to .NET Core as easy as possible as well as other changes that will help in that.

**Comment by erinxocon** [comment]
@petrroll 

Good Advice,  I'm just anxious to get this library running in some linux and mac applications I'm developing. 🙂

**Comment by petrroll** [comment]
@erinxocon I understand that :). At the same time looking at the portability analysis linked in different issue it seems to me that the only thing blocking easy porting to .NET Core is HttpClient/WebClient which isn't a problem directly linked to new HttpServer.

I'm all for switching to WebClient, though.

---
Also, I believe that it would be greatly beneficial to split the API project into two, one for WebAPI that is platform agnostic (is portable to .NET Core) and LocalAPI that is strictly desktop Windows only. Afterall the WebAPI could be useful for WebServers, Linux/MacOSX apps, Xamarin Apps... you name it.

**Comment by erinxocon** [comment]
> Afterall the WebAPI could be useful for WebServers, Linux/MacOSX apps, Xamarin Apps... you name it.

Exactly!  When I did my initial build using dotnetcore I'm down to only the issues with the HTTPClient/Server.  Splitting the project to make it platform agnostic has it's benefits in my opinion.  

@JohnnyCrazy  Thoughts?

**Comment by petrroll** [comment]
@erinxocon  There's an initial [split PR now](https://github.com/JohnnyCrazy/SpotifyAPI-NET/pull/102). Will look into the HttpClient/WebClient issue soon but now I really have to go to get some sleep (a bit late here in central Europe :))
